File: apps/web/src/lib/client/region-store.test.ts
import { beforeEach, describe, expect, it } from "vitest";

import {
  moveRegion,
  addRegion,
  createEmptyRegionStore,
  loadRegionStore,
  migrateRegionStore,
  REGION_STORE_KEY,
  removeRegion,
  selectRegion,
  setConsent,
  type StoredRegion,
} from "./region-store";

const NONSAN: StoredRegion = { sigunCode: "44230", facCode: "4423010045" };
const JEJU: StoredRegion = { sigunCode: "50110", facCode: "5011010001" };

beforeEach(() => {
  window.localStorage.clear();
});

describe("region-store 저장·선택·삭제", () => {
  it("빈 localStorage에서는 안전한 초기 상태를 돌려준다", () => {
    expect(loadRegionStore()).toEqual({
      schemaVersion: 1,
      consentVersion: null,
      regions: [],
      currentIndex: 0,
    });
  });

  it("지역을 추가하면 mulsigye:v1 키에 저장되고 현재 지역이 된다", () => {
    addRegion(NONSAN);
    const store = addRegion(JEJU);

    expect(store.regions).toEqual([NONSAN, JEJU]);
    expect(store.currentIndex).toBe(1);
    expect(window.localStorage.getItem(REGION_STORE_KEY)).not.toBeNull();
    expect(loadRegionStore()).toEqual(store);
  });

  it("같은 시군을 다시 추가하면 중복 없이 그 지역을 선택한다", () => {
    addRegion(NONSAN);
    addRegion(JEJU);
    const store = addRegion(NONSAN);

    expect(store.regions).toHaveLength(2);
    expect(store.currentIndex).toBe(0);
  });

  it("selectRegion은 currentIndex를 저장하고 범위 밖 값은 보정한다", () => {
    addRegion(NONSAN);
    addRegion(JEJU);

    expect(selectRegion(0).currentIndex).toBe(0);
    expect(loadRegionStore().currentIndex).toBe(0);
    expect(selectRegion(99).currentIndex).toBe(1);
    expect(selectRegion(-3).currentIndex).toBe(0);
  });

  it("현재 지역을 삭제하면 currentIndex를 남은 범위로 보정한다", () => {
    addRegion(NONSAN);
    addRegion(JEJU);
    selectRegion(1);

    const store = removeRegion(JEJU.sigunCode);

    expect(store.regions).toEqual([NONSAN]);
    expect(store.currentIndex).toBe(0);
    expect(loadRegionStore()).toEqual(store);
  });

  it("앞쪽 지역을 삭제하면 현재 지역 선택이 유지된다", () => {
    addRegion(NONSAN);
    addRegion(JEJU);
    selectRegion(1);

    const store = removeRegion(NONSAN.sigunCode);

    expect(store.regions).toEqual([JEJU]);
    expect(store.currentIndex).toBe(0);
  });

  it("setConsent는 동의 버전만 갱신한다", () => {
    addRegion(NONSAN);
    const store = setConsent("consent-v1");

    expect(store.consentVersion).toBe("consent-v1");
    expect(store.regions).toEqual([NONSAN]);
    expect(loadRegionStore().consentVersion).toBe("consent-v1");
  });
});

describe("region-store 안전 초기화·마이그레이션 훅", () => {
  it("손상된 JSON이면 예외 없이 초기 상태로 시작한다", () => {
    window.localStorage.setItem(REGION_STORE_KEY, "{잘못된 json");

    expect(loadRegionStore()).toEqual(createEmptyRegionStore());
  });

  it("알 수 없는 schemaVersion이면 초기 상태로 시작한다", () => {
    window.localStorage.setItem(
      REGION_STORE_KEY,
      JSON.stringify({ schemaVersion: 99, regions: [] }),
    );

    expect(loadRegionStore()).toEqual(createEmptyRegionStore());
  });

  it("마이그레이션 훅은 현재 버전 페이로드를 정규화해 통과시킨다", () => {
    const migrated = migrateRegionStore({
      schemaVersion: 1,
      consentVersion: "consent-v1",
      regions: [NONSAN],
      currentIndex: 5,
    });

    expect(migrated).toEqual({
      schemaVersion: 1,
      consentVersion: "consent-v1",
      regions: [NONSAN],
      currentIndex: 0,
    });
  });

  it("마이그레이션 훅은 알 수 없는 버전·형태에 null을 돌려준다", () => {
    expect(migrateRegionStore({ schemaVersion: 2, regions: [] })).toBeNull();
    expect(
      migrateRegionStore({ schemaVersion: 1, regions: "없음" }),
    ).toBeNull();
    expect(migrateRegionStore("문자열")).toBeNull();
  });
});

describe("region-store 개인정보 최소화", () => {
  it("주소 원문·지역 이름은 어떤 키에도 저장되지 않는다", () => {
    const leaky = {
      ...NONSAN,
      label: "충남 논산시 시민로 210",
      address: "충남 논산시",
      sigunName: "논산시",
      query: "논산 시민로",
    } as unknown as StoredRegion;

    addRegion(leaky);
    setConsent("consent-v1");

    const raw = window.localStorage.getItem(REGION_STORE_KEY);
    expect(raw).not.toBeNull();
    expect(raw).not.toContain("논산");
    expect(raw).not.toContain("시민로");

    const parsed = JSON.parse(raw as string) as Record<string, unknown>;
    expect(Object.keys(parsed).sort()).toEqual([
      "consentVersion",
      "currentIndex",
      "regions",
      "schemaVersion",
    ]);
    const regions = parsed.regions as Record<string, unknown>[];
    expect(regions).toHaveLength(1);
    expect(Object.keys(regions[0] as object).sort()).toEqual([
      "facCode",
      "sigunCode",
    ]);
  });

  it("mulsigye:v1 밖의 다른 키를 만들지 않는다", () => {
    addRegion(NONSAN);
    selectRegion(0);
    setConsent("consent-v1");

    expect(window.localStorage.length).toBe(1);
    expect(window.localStorage.key(0)).toBe(REGION_STORE_KEY);
  });
});

// 드래그의 접근성 대체(위로/아래로 이동)와 같은 연산이다. 맨 위가 기본 주소지이므로
// 순서 변경은 곧 대표 지역 변경이기도 하다.
describe("moveRegion", () => {
  const SANCHEONG: StoredRegion = {
    sigunCode: "48860",
    facCode: "4886010001",
  };

  function seedThree() {
    addRegion(NONSAN);
    addRegion(JEJU);
    addRegion(SANCHEONG);
  }

  it("항목을 위로 올린다", () => {
    seedThree();

    const store = moveRegion(2, 1);

    expect(store.regions.map((region) => region.sigunCode)).toEqual([
      NONSAN.sigunCode,
      SANCHEONG.sigunCode,
      JEJU.sigunCode,
    ]);
  });

  it("선택은 칸이 아니라 지역을 따라간다", () => {
    seedThree();
    selectRegion(0); // 논산을 보고 있다

    const store = moveRegion(2, 0);

    expect(store.regions[0]?.sigunCode).toBe(SANCHEONG.sigunCode);
    // 논산이 한 칸 밀렸으니 선택 인덱스도 함께 밀려야 계속 논산을 본다.
    expect(store.currentIndex).toBe(1);
  });

  it("범위를 벗어난 이동은 끝으로 자른다", () => {
    seedThree();

    const store = moveRegion(0, -3);

    expect(store.regions[0]?.sigunCode).toBe(NONSAN.sigunCode);
  });

  it("같은 자리로 옮기면 아무 것도 바뀌지 않는다", () => {
    seedThree();
    const before = loadRegionStore();

    const store = moveRegion(1, 1);

    expect(store.regions).toEqual(before.regions);
    expect(store.currentIndex).toBe(before.currentIndex);
  });

  it("빈 목록에서도 안전하다", () => {
    const store = moveRegion(0, 1);

    expect(store.regions).toEqual([]);
    expect(store.currentIndex).toBe(0);
  });
});

File: apps/android/app/src/test/kotlin/com/mulsigye/app/core/storage/RegionStoreTest.kt
package com.mulsigye.app.core.storage

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.PreferenceDataStoreFactory
import androidx.datastore.preferences.core.edit
import java.io.File
import java.nio.file.Files
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.TestScope
import kotlinx.coroutines.test.UnconfinedTestDispatcher
import kotlinx.coroutines.test.runTest
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@OptIn(ExperimentalCoroutinesApi::class)
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class RegionStoreTest {
    private val testScope = TestScope(UnconfinedTestDispatcher())
    private lateinit var tmpDir: File
    private lateinit var dataStore: DataStore<Preferences>
    private lateinit var store: RegionStore

    private val nonsan = StoredRegion(sigunCode = "44230", facCode = "4423010045")
    private val naju = StoredRegion(sigunCode = "46170", facCode = "4617010001")
    private val jeju = StoredRegion(sigunCode = "50110", facCode = "5011010004")

    @Before
    fun setUp() {
        tmpDir = Files.createTempDirectory("regionstore").toFile()
        dataStore = PreferenceDataStoreFactory.create(scope = testScope.backgroundScope) {
            File(tmpDir, "region.preferences_pb")
        }
        store = RegionStore(dataStore)
    }

    @After
    fun tearDown() {
        tmpDir.deleteRecursively()
    }

    @Test
    fun startsEmpty() = testScope.runTest {
        val state = store.regionStoreFlow.first()
        assertEquals(RegionStore.SCHEMA_VERSION, state.schemaVersion)
        assertNull(state.consentVersion)
        assertTrue(state.regions.isEmpty())
        assertEquals(0, state.currentIndex)
    }

    @Test
    fun addsRegionsAndSelectsTheNewest() = testScope.runTest {
        store.addRegion(nonsan)
        store.addRegion(naju)
        val state = store.regionStoreFlow.first()
        assertEquals(listOf(nonsan, naju), state.regions)
        assertEquals(1, state.currentIndex)
    }

    @Test
    fun addingSameSigunDeduplicatesAndSelectsIt() = testScope.runTest {
        store.addRegion(nonsan)
        store.addRegion(naju)
        store.addRegion(StoredRegion(sigunCode = "44230", facCode = "4423099999"))
        val state = store.regionStoreFlow.first()
        assertEquals(2, state.regions.size)
        assertEquals("4423099999", state.regions[0].facCode)
        assertEquals(0, state.currentIndex)
    }

    @Test
    fun selectRegionClampsOutOfRange() = testScope.runTest {
        store.addRegion(nonsan)
        store.addRegion(naju)
        store.selectRegion(9)
        assertEquals(1, store.regionStoreFlow.first().currentIndex)
        store.selectRegion(-3)
        assertEquals(0, store.regionStoreFlow.first().currentIndex)
    }

    @Test
    fun removingRegionBeforeCurrentShiftsIndex() = testScope.runTest {
        store.addRegion(nonsan)
        store.addRegion(naju)
        store.addRegion(jeju) // currentIndex = 2
        store.removeRegion("44230") // remove index 0
        val state = store.regionStoreFlow.first()
        assertEquals(listOf(naju, jeju), state.regions)
        assertEquals(1, state.currentIndex)
    }

    @Test
    fun removingCurrentRegionClampsIndex() = testScope.runTest {
        store.addRegion(nonsan)
        store.addRegion(naju) // currentIndex = 1
        store.removeRegion("46170") // remove the selected last one
        val state = store.regionStoreFlow.first()
        assertEquals(listOf(nonsan), state.regions)
        assertEquals(0, state.currentIndex)
    }

    @Test
    fun setConsentPersistsOnlyTheVersion() = testScope.runTest {
        store.setConsent("consent-v1")
        assertEquals("consent-v1", store.regionStoreFlow.first().consentVersion)
    }

    @Test
    fun unknownSchemaVersionResetsToEmpty() = testScope.runTest {
        dataStore.edit {
            it[RegionStore.KEY] =
                """{"schemaVersion":2,"consentVersion":"x","regions":[{"sigunCode":"44230","facCode":"4423010045"}],"currentIndex":0}"""
        }
        val state = store.regionStoreFlow.first()
        assertTrue(state.regions.isEmpty())
        assertNull(state.consentVersion)
    }

    @Test
    fun corruptJsonResetsToEmpty() = testScope.runTest {
        dataStore.edit { it[RegionStore.KEY] = "{ this is not json" }
        assertTrue(store.regionStoreFlow.first().regions.isEmpty())
    }

    @Test
    fun currentIndexBeyondRangeIsClampedOnRead() = testScope.runTest {
        dataStore.edit {
            it[RegionStore.KEY] =
                """{"schemaVersion":1,"consentVersion":null,"regions":[{"sigunCode":"44230","facCode":"4423010045"}],"currentIndex":7}"""
        }
        assertEquals(0, store.regionStoreFlow.first().currentIndex)
    }

    @Test
    fun legacyPayloadWithoutRegionsKeepsHasEverRegisteredFalse() = testScope.runTest {
        // 지역이 없는 구버전 payload는 최초 사용자로 본다(false 유지).
        dataStore.edit {
            it[RegionStore.KEY] =
                """{"schemaVersion":1,"consentVersion":"consent-v1","regions":[],"currentIndex":0}"""
        }
        assertFalse(store.regionStoreFlow.first().hasEverRegistered)
    }

    @Test
    fun neverStoresAddressOrName() = testScope.runTest {
        // 주소·이름이 섞여 들어온 저장값을 읽어도 코드 2종만 남기고 재저장한다.
        dataStore.edit {
            it[RegionStore.KEY] =
                """{"schemaVersion":1,"consentVersion":null,"regions":[{"sigunCode":"44230","facCode":"4423010045","address":"전라남도 나주시 시청길 22","name":"논산시"}],"currentIndex":0}"""
        }
        store.selectRegion(0) // 읽기→변환→재저장 경유

        val raw = dataStore.data.first()[RegionStore.KEY]!!
        assertTrue(raw.contains("44230"))
        assertTrue(raw.contains("4423010045"))
        assertFalse(raw.contains("address"))
        assertFalse(raw.contains("시청길"))
        assertFalse(raw.contains("논산시"))
    }

    @Test
    fun moveRegionReordersAndKeepsCurrentPointingAtSameRegion() = testScope.runTest {
        store.addRegion(nonsan)
        store.addRegion(naju)
        store.addRegion(jeju) // regions = [nonsan, naju, jeju], currentIndex = 2 (jeju)

        // 맨 위(0)를 맨 아래(2)로 이동 → currentIndex는 여전히 jeju를 가리켜야 한다.
        store.moveRegion(0, 2)
        var state = store.regionStoreFlow.first()
        assertEquals(listOf(naju, jeju, nonsan), state.regions)
        assertEquals(1, state.currentIndex) // jeju는 이제 index 1

        // 현재 선택(jeju, index 1)을 맨 위로 이동 → currentIndex가 0으로 따라간다.
        store.moveRegion(1, 0)
        state = store.regionStoreFlow.first()
        assertEquals(listOf(jeju, naju, nonsan), state.regions)
        assertEquals(0, state.currentIndex)
    }

    @Test
    fun moveRegionIgnoresOutOfRangeOrNoop() = testScope.runTest {
        store.addRegion(nonsan)
        store.addRegion(naju)
        store.moveRegion(0, 0) // no-op
        store.moveRegion(0, 5) // out of range
        store.moveRegion(-1, 1) // out of range
        val state = store.regionStoreFlow.first()
        assertEquals(listOf(nonsan, naju), state.regions)
    }

    @Test
    fun hasEverRegisteredStartsFalseAndFlipsTrueOnAdd() = testScope.runTest {
        assertFalse(store.regionStoreFlow.first().hasEverRegistered)
        store.addRegion(nonsan)
        assertTrue(store.regionStoreFlow.first().hasEverRegistered)
    }

    @Test
    fun hasEverRegisteredStaysTrueAfterRemovingAllRegions() = testScope.runTest {
        store.addRegion(nonsan)
        store.removeRegion("44230")
        val state = store.regionStoreFlow.first()
        assertTrue(state.regions.isEmpty())
        assertTrue(state.hasEverRegistered) // 재방문(지역 삭제) 사용자 구분용으로 유지된다.
    }

    @Test
    fun hasEverRegisteredPersistsInStoredPayload() = testScope.runTest {
        store.addRegion(nonsan)
        val raw = dataStore.data.first()[RegionStore.KEY]!!
        assertTrue(raw.contains("hasEverRegistered"))
        assertTrue(raw.contains("true"))
    }

    @Test
    fun decodeMigratesHasEverRegisteredForOldPayloadWithRegions() = testScope.runTest {
        // 구 페이로드(필드 없음)라도 지역이 있으면 등록 이력이 있는 것이므로 true로 마이그레이션한다
        // (재방문 사용자가 지역을 지웠을 때 최초 검색으로 튀지 않도록).
        dataStore.edit {
            it[RegionStore.KEY] =
                """{"schemaVersion":1,"consentVersion":"consent-v1","regions":[{"sigunCode":"44230","facCode":"4423010045"}],"currentIndex":0}"""
        }
        val state = store.regionStoreFlow.first()
        assertTrue(state.hasEverRegistered)
        assertEquals(1, state.regions.size) // 나머지 필드는 정상 복원.
        assertEquals("consent-v1", state.consentVersion)
    }

    @Test
    fun addRegionWritesOnlyTwoCodes() = testScope.runTest {
        store.addRegion(nonsan)
        val raw = dataStore.data.first()[RegionStore.KEY]!!
        assertTrue(raw.contains("sigunCode"))
        assertTrue(raw.contains("facCode"))
        assertFalse(raw.contains("address"))
        assertFalse(raw.contains("label"))
    }
}
